osier/models/deap_runner.py

```diff
@@ -0,0 +1,247 @@
+import numpy as np
+from pathlib import Path
+import dill
+import random
+import time
+
+
+from pymoo.core.problem import ElementwiseProblem, Problem
+from pymoo.util.ref_dirs import get_reference_directions
+from deap import algorithms
+from deap import base
+from deap.benchmarks.tools import igd
+from deap import creator
+from deap import tools
+
+from osier import *
+from osier.tech_library import *
+
+
+def uniform(low, up, size=None):
+    try:
+        return [random.uniform(a, b) for a, b in zip(low, up)]
+    except TypeError:
+        return [random.uniform(a, b) for a, b in zip([low] * size, [up] * size)]
+
+
+algorithm_options = {'nsga2':tools.selNSGA2,
+                     'nsga3':tools.selNSGA3}
+
+class OsierDEAP(object):
+    """
+    A DEAP algorithm object. Holds information and parameters to
+    run a genetic algorithm using DEAP.
+
+    Parameters
+    ----------
+    problem : :class:`Osier.CapacityExpansion`
+        The problem to execute. In theory, this could be any object
+        that has a method :meth:`evaluate` (for example, if it inherits 
+        from :class:`pymoo.ElementwiseProblem`) and outputs a dictionary.
+    algorithm : Optional, str
+        Indicates the algorithm to use. Accepts ['nsga2', 'nsga3']. 
+        NSGA2 is recommended for problems with three or fewer objectives.
+        NSGA3 is recommended for problems with three or more objectives. 
+    lower_bound : Optional, float, int
+        The lower bound for each 'gene' in a candidate individual.
+    upper_bound : Optional, float, int
+        The upper bound for each 'gene' in a candidate individual.
+    repair : Optional, :class:`pymoo.Repair`
+        Adjusts the candidate individuals based on a user-defined procedure.
+    hyper_params : Optional, dict
+        Contains hyperparameters for the selection algorithm.
+    save_directory : Optional, str
+        Indicates the read/write directory.
+    """
+    def __init__(self, 
+                 problem,
+                 algorithm=None, 
+                 lower_bound=0.0,
+                 upper_bound=None,
+                 repair=None,
+                 hyper_params=None,
+                 pop_size=100,
+                 save_directory=None
+                 ) -> None:
+        self.problem = problem
+        self.n_obj = self.problem.n_obj
+        if algorithm:
+            self.algorithm = algorithm.lower()
+        elif self.n_obj < 3:
+            self.algorithm = 'nsga2'
+        elif self.n_obj >= 3:
+            self.algorithm = 'nsga3' 
+        self.n_dim = self.problem.n_var
+        self.pop_size = pop_size
+        self.repair = repair
+        self.completed_generations = 0
+        self.solve_time = 0.0
+        self.last_population = None
+        self.save_directory = Path(save_directory) if save_directory else None
```

Comment:
`save_directory` default appears to be `f"{timestr}-OsierModel.pkl"`, how come you don't define it here instead?

Reply:
The default is `None`. The file NAME is `f"{timestr}-OsierModel.pkl"`. It either saves the file in the current directory or a specific directory.